java/client/src/org/openqa/selenium/remote/RemoteWebDriverBuilder.java

```diff
@@ -174,261 +191,260 @@ public RemoteWebDriverBuilder url(String url) {
-  }
+    remoteHost = uri;
 
-  private Map<String, Object> validate(Capabilities options) {
-    return options.asMap().entrySet().stream()
-        // Ensure that the keys are ok
-        .peek(
-            entry -> {
-              if (!OK_KEYS.test(entry.getKey())) {
-                throw new IllegalArgumentException(
-                    "Capability key is not a valid w3c key: " + entry.getKey());
-              }
-            })
-        // And remove null values, as these are ignored.
-        .filter(entry -> entry.getValue() != null)
-        .collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+    return this;
   }
 
-  @VisibleForTesting
-  Plan getPlan() {
-    return new Plan();
+  /**
+   * Allows precise control of the {@link ClientConfig} to use with remote
+   * instances. If {@link ClientConfig#baseUri(URI)} has been called, then
+   * that will be used as the base URI for the session.
+   */
+  public RemoteWebDriverBuilder config(ClientConfig config) {
+    Require.nonNull("HTTP client config", config);
+
+    if (config.baseUri() != null) {
+      if (remoteHost != null || driverService != null) {
+        throw new IllegalArgumentException("Base URI has already been set. Cannot also set it via client config");
+      }
+    }
+
+    this.clientConfig = config;
+
+    return this;
   }
 
-  private void validateDriverServiceAndUrlConstraint() {
-    if (remoteHost != null && service != null) {
-      throw new IllegalArgumentException(
-          "You may only set one of the remote url or the driver service to use.");
+  /**
+   * Use the given {@link DriverService} to set up the webdriver instance. It is an error to set
+   * both this and also call {@link #url(URL)}.
+   */
+  public RemoteWebDriverBuilder withDriverService(DriverService service) {
+    Require.nonNull("Driver service", service);
+
+    if (clientConfig.baseUri() != null || remoteHost != null) {
+      throw new IllegalArgumentException("Base URI has already been set. Cannot also set driver service.");
     }
+
+    this.driverService = service;
+
+    return this;
   }
 
   @VisibleForTesting
-  class Plan {
+  RemoteWebDriverBuilder connectingWith(Function<ClientConfig, HttpHandler> handlerFactory) {
+    Require.nonNull("Handler factory", handlerFactory);
+    this.handlerFactory = handlerFactory;
+    return this;
+  }
 
-    private Plan() {
-      // Not for public consumption
+  @VisibleForTesting
+  WebDriver getLocalDriver() {
+    if (remoteHost != null || clientConfig.baseUri() != null || driverService != null) {
+      return null;
     }
 
-    boolean isUsingDriverService() {
-      return remoteHost == null;
+    Set<WebDriverInfo> infos = StreamSupport.stream(
+         ServiceLoader.load(WebDriverInfo.class).spliterator(),
+         false)
+         .filter(WebDriverInfo::isAvailable)
+         .collect(Collectors.toSet());
+
+    Capabilities additional = new ImmutableCapabilities(additionalCapabilities);
```

Comment:
Is there a chance someone has caps for EdgeHTML and the driver for ChromiumEdge is returned?

Reply:
We filter by whether or not the `WebDriverInfo` says the driver is available, and whether that `WebDriverInfo` supports the `Capabilities`. Additionally, we return a factory that captures the `WebDriverInfo` and the `Capabilities` it matched against.

That's a long way of saying that if a user asked for EdgeHTML and got Chromium, there's a bug in the `WebDriverInfo`